### DISCUSSION TITLE
The Conv2d layer: input and output traits

### DISCUSSION THREAD

**Discussion by lodykas** [neuronika/neuronika#83] [Q&A] [answered]
Hello,

I've been playing around with this crate, and it's been the simplest and most intuitive of its kind (In my humble opinion) that I've found so far.

That is, until I tried using Conv2d layers. I'll first explain what i tried and then what I think my problem is. (Now, keep in mind, playing around with this crate is the second project I used to learn Rust, so it might as well just be my mistake)

Since I had a lot of success doing a full (albeit small) Classifier pipeline on a toy dataset with Linear layers, I tried the next step, which is of course, a Neural Network with a dynamic number of Conv2d layers followed by another number of Linear layers, in the likes of:
``` Rust

fn new(nb_conv:usize, nb_lin: usize, width:usize, height:usize, pad:Pad) -> Self {
	let mut status = ModelStatus::default();
	
	let con_layers = Vec::with_capacity(nb_conv);
	for _ in 0..nb_conv{
		con_layers.push(status.register(
			Conv2d::new(5, 5, (3, 3), (0, 0), pad, (1, 1), (1, 1))
		));
	}

	let lin_layers = Vec::with_capacity(nb_lin);
	for _ in 0..nb_lin{
		lin_layers.push(status.register(
			Linear::new(5, 5)
		));
	}
	
	Self {
		con_layers: con_layers,
		lin_layers: lin_layers,
		status,
	}
}
```

There are many problems with this, most of which I would "figure out later", like the matching of dimensions of the tensors between each layer, But the main issue I had was with the forward function.

This would be the basic idea:
``` Rust
let mut out = input;
for layer in self.con_layers{
	out = layer.forward(out);
}
```

but even going down to something like
``` Rust
let out = self.con1.forward(input);
let out1 = self.con2.forward(out);
```
Won't compile because of Traits condition on the Conv2d layer.

Now, for the actual question:
The input has a trait bound of Convolve<>, itself conditioned on some Data + Display + Debug traits; but the output offers no garantees of the Debug and Display Traits, hence it won't compile. Why do the input and output differs in Trait bounds?
I'd be very interested in a example of usage of those Conv layers, as I couldn't find any to compare and find out what I did wrong.

I'd really hate it if I had to change to another crate, since I had a lot of fun with this crate so far... Anyways, I can show my test code to showcase the error if needed, but for now I'll keep it 'short'

Thanks

**Comment by frjnn** [comment, marked as answer]
Hi, first and foremost, thank you for spotting this and for using this project. 

You are right indeed, such bounds should not be there, I must have messed up while implementing the debug and display traits for other nodes, apologies.

If its in your possibilities I would ask you to provide the full code so that I can be 100% positive to have fixed the issue.

**Reply by lodykas** [reply]
``` Rust
use ndarray::{Ix4, Dim};

use neuronika::Param;
use neuronika::{Backward, Data, Forward, Gradient, MatMatMulT, Overwrite, VarDiff, Convolve};
use neuronika::nn::{ModelStatus, Conv2d, Learnable};
use neuronika::nn;

// Network definition.
struct NeuralNetwork<Pad: nn::PaddingMode> {
	con1: Conv2d<Pad>,
	con2: Conv2d<Pad>,
	con3: Conv2d<Pad>,
	con4: Conv2d<Pad>,
	con5: Conv2d<Pad>,
	status: ModelStatus,     
}

impl<Pad: nn::PaddingMode> NeuralNetwork<Pad> {
	fn new(nb_conv:usize, nb_lin: usize, width:usize, height:usize, pad:Pad) -> Self {
		let mut status = ModelStatus::default();

		let con1 = status.register(
				Conv2d::new(5, 5, (3, 3), (0, 0), pad, (1, 1), (1, 1))
		);
		let con2 = status.register(
			Conv2d::new(5, 5, (3, 3), (0, 0), pad, (1, 1), (1, 1))
		);
		let con3 = status.register(
			Conv2d::new(5, 5, (3, 3), (0, 0), pad, (1, 1), (1, 1))
		);
		let con4 = status.register(
			Conv2d::new(5, 5, (5, 4), (0, 0), pad, (1, 1), (1, 1))
		);
		let con5 = status.register(
			Conv2d::new(5, 4, (1, 1), (0, 0), pad, (1, 1), (1, 1))
		);
		
		Self {
			con1: con1,
			con2: con2,
			con3: con3,
			con4: con4,
			con5: con5,
			status,
		}
	}

    fn forward<I, T, U>(
        &self,
        input: I,
    ) -> VarDiff<impl Data<Dim = Ix4> + Forward, impl Gradient<Dim = Ix4> + Overwrite + Backward>
    where
		I: Convolve<I, Learnable<Ix4>, Pad>,
		I::Output: Into<VarDiff<T, U>>,
		T: Data<Dim = Ix4>,
		U: Gradient<Dim = Ix4> + Overwrite,
    {
		let out = self.con1.forward(input);
		let out1 = self.con2.forward(out);
		let out2 = self.con3.forward(out1);
		let out3 = self.con4.forward(out2);
		let out4 = self.con5.forward(out3);
		
		out4
    }


	fn parameters(&self) -> Vec<Param> {
		self.status.parameters()
	}

	/// Switches the network in training mode.
	fn train(&self) {
		self.status.train()
	}

	/// Switches the network in inference mode.
	fn eval(&self) {
		self.status.eval()
	}
}
```

Here it is. it was not yet instanciated since I couldn't compile it, so not much more yet

**Reply by frjnn** [reply]
Hi, thanks again for pointing out this issue, your code should now compile fine. Let me know if you encounter any more problems. Your help is greatly appreciated.

**Comment by ducklin5** [comment]
@lodykas Seems like you were also trying to pass the output of a Conv2D layer to a Linear Layer. Did you manage to get this working?
@frjnn  Is there some way to forward a conv2d output to a linear layer while allowing the result to be differentiable?

Just passing the output in does not work because they use different array sizes
```rust
let out2 = self.conv2.forward(out1);
let out3 = self.lin1.forward(out2);
```
Gives this message on the forward call of the 2nd line:
```
type mismatch resolving `<impl neuronika::Data<Dim = ndarray::Dim<[usize; 4]>> + neuronika::Forward as neuronika::Data>::Dim == ndarray::Dim<[usize; 2]>`

expected an array with a fixed size of 2 elements, found one with 4 elements
```

**Reply by frjnn** [reply]
You will need a differentiable `reshape()` method, which we current don't have but it'd be super easy to implement. 

We are currently reworking the whole API , see [here](https://github.com/neuronika/neuronika/tree/variable-rework) if you want to check, and I can add it there.

**Reply by ducklin5** [reply]
That would be awesome! I believe PyTorch has something similar: https://pytorch.org/docs/stable/tensor_view.html.
I'd have implemented a trait if I knew how to. I've very new to both ML and Rust.

**Reply by lodykas** [reply]
What I did was applying a conv layer with a single filter, this filter having the exact same size as the input. This should be equivalent to a linear layer, the only dimensions that change are the input and output channels. 

However, this is annoying because you need to interpret a BX1X1XC matrix as a BXC vector (B: Batch size and C: Channel count), so a  reshape operation would be welcome indeed, but if you can't wait there's that
